### PR TITLE
[Merged by Bors] - Immutable sparse sets for metadata storage

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Make core types in ECS smaller. The column sparse set in Tables is never updated after creation.

## Solution
Create `ImmutableSparseSet` which removes the capacity fields in the backing vec's and the APIs for inserting or removing elements. Drops the size of the sparse set by 3 usizes (24 bytes on 64-bit systems)

## Followup
~~After #4809, Archetype's component SparseSet should be replaced with it.~~ This has been done.

---

## Changelog
Removed: `Table::component_capacity`

## Migration Guide
`Table::component_capacity()` has been removed as Tables do not support adding/removing columns after construction.

### CODE REVIEW

File: crates/bevy_ecs/src/storage/sparse_set.rs [hunk 1]
```diff
@@ -89,6 +89,33 @@ impl<I: SparseSetIndex, V> SparseArray<I, V> {
...
+    pub fn to_immutable(self) -> ImmutableSparseArray<I, V> {
+        ImmutableSparseArray {
+            values: self.values.into_boxed_slice(),
+            marker: PhantomData,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ImmutableSparseArray<I, V = I> {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/sparse_set.rs:102]: Basic doc strings please.

File: crates/bevy_ecs/src/storage/sparse_set.rs [hunk 2]
```diff
@@ -426,6 +453,73 @@ impl<I: SparseSetIndex, V> SparseSet<I, V> {
...
+        ImmutableSparseSet {
+            dense: self.dense.into_boxed_slice(),
+            indices: self.indices.into_boxed_slice(),
+            sparse: self.sparse.to_immutable(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ImmutableSparseSet<I, V: 'static> {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/sparse_set.rs:467]: Needs basic docs.

File: crates/bevy_ecs/src/storage/table.rs [hunk 3]
```diff
@@ -186,38 +186,45 @@ impl Column {
     }
 }
 
-pub struct Table {
+pub struct TableBuilder {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/table.rs:189]: Basic docs.

File: crates/bevy_ecs/src/storage/table.rs [hunk 4]
```diff
@@ -186,38 +186,45 @@ impl Column {
...
+
+    pub fn build(self) -> Table {
+        Table {
+            columns: self.columns.to_immutable(),
+            entities: Vec::with_capacity(self.capacity),
         }
     }
+}
+
+pub struct Table {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/table.rs:217]: Definitely needs docs; this is an deceptively simple public type name.

File: crates/bevy_ecs/src/storage/table.rs [hunk 5]
```diff
@@ -186,38 +186,45 @@ impl Column {
...
+pub struct TableBuilder {
     columns: SparseSet<ComponentId, Column>,
-    entities: Vec<Entity>,
+    capacity: usize,
 }
 
-impl Table {
-    pub const fn new() -> Table {
+impl TableBuilder {
+    pub fn with_capacity(capacity: usize, column_capacity: usize) -> Self {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/table.rs:195]: Can you document the difference between the two fields used here?

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
The strategy makes sense to me, but this needs docs. The internals are slowly getting less intimidating, let's not reverse that trend.

**Review by alice-i-cecile — APPROVED** [review]
I'm happy with this now. I think it's unlikely that we ever will ever want to remove columns (i.e. registered resource / component types), even once we start doing empty archetype cleanup. And if we do, this is a more fair comparison for benchmarks.

File: crates/bevy_ecs/src/storage/sparse_set.rs [hunk 6]
```diff
@@ -14,6 +14,14 @@ pub(crate) struct SparseArray<I, V = I> {
     marker: PhantomData<I>,
 }
 
+/// A space-optimized version of [`SparseArray`] that cannot be changed
+/// after contrusction.
```
Comment (kristoff3r) [inline crates/bevy_ecs/src/storage/sparse_set.rs:18]:
```suggestion
/// after construction.
```

File: crates/bevy_ecs/src/storage/sparse_set.rs [hunk 7]
```diff
@@ -249,11 +273,71 @@ pub struct SparseSet<I, V: 'static> {
     sparse: SparseArray<I, usize>,
 }
 
+/// A space-optimized version of [`SparseSet`] that cannot be changed
+/// after contrusction.
```
Comment (kristoff3r) [inline crates/bevy_ecs/src/storage/sparse_set.rs:277]:
```suggestion
/// after construction.
```

**Review by Diddykonga — APPROVED** [review]
LGTM
Don't personally like simple macro usage, but I digress.

File: crates/bevy_ecs/src/storage/table.rs [hunk 8]
```diff
@@ -262,31 +262,68 @@ impl Column {
...
-    entities: Vec<Entity>,
+    capacity: usize,
 }
 
-impl Table {
-    pub(crate) fn with_capacity(capacity: usize, column_capacity: usize) -> Table {
+impl TableBuilder {
+    /// Creates a blank [`Table`], allocating space for `column_capacity` columns
+    /// with the capacity to hold `capacity` entities worth of components each.
+    pub fn with_capacity(capacity: usize, column_capacity: usize) -> Self {
```
Comment (cart) [inline crates/bevy_ecs/src/storage/table.rs:282]: Is there a chance that we end up realloc-ing our tables on construction in some / all cases, due to the fact that with_capacity isn't guaranteed to allocate exactly `capacity`, but `shrink_to_fit` will shrink to capacity?

Comment (james7132) [reply crates/bevy_ecs/src/storage/table.rs:282]: There is, but unless we're constantly making new `Tables` and `Archetypes`, the cost IMO is negligible long term.

Comment (cart) [reply crates/bevy_ecs/src/storage/table.rs:282]: The frequency might go up drastically if we start doing "fragmenting relations". In that context the bytes saved might not be worth it. Hard to say what to value more.

I guess if/when we implement fragmenting relations we can let benchmarks decide. But this level of nuance is easy to forget about.

**Comment by cart** [comment]
Probably going to conflict with #4927 but lets give it a try

**Pull request closed without merging**